### ISSUE TITLE
Implement data extraction via web browser (Chrome devtools protocol)

### ISSUE THREAD

**Issue by bastimeyer** [streamlink/streamlink#5380]
Meta-thread for the implementation of data extraction via a web browser.

----

## Motivation

We need a way to extract data from websites which make it impossible to do this in a pure Python implementation, e.g. because of JS runtime and DOM requirements, etc.

There are Python implementations for executing JS code, and other projects also try to execute JS code in NodeJS, but none of this solves the problem we're facing with the Twitch plugin, not to mention that running untrusted JS code in NodeJS is insanely stupid.

See the following thread/comment which explains this in detail:

- https://github.com/streamlink/streamlink/issues/5370#issuecomment-1582806603

## Rationale

In the linked comment I came to the conclusion that adding `selenium` as a dependency for this task is impossible, because of the detection methods which certain sites implement when the web browser is launched via its `webdriver` interface, which is what Selenium does. It's not just that `Navigator.prototype.webdriver` returns `true` in that case, but there are many other ways for detecting this different environment the browser is run in.

This is a bit unfortunate, because the W3C `webdriver` and `bi-directional webdriver` specifications are the correct way to automate browsers, and it sadly limits us to chromium-based browsers and an implementation based on the Chrome Devtools Protocol (CDP) for remotely and bi-directionally controlling the browser process. The reason for this is that Chromium works just like normal when it's launched with the `--remote-debugging-port=...` parameter (unless the port is set to 0 for some reason) compared to using `chromedriver` (which itself is communicating via the CDP). Firefox also implements the CDP, but only a tiny subset which is unsuitable for our needs. Safari doesn't support this and other non-Chromium browsers are mostly irrelevant.

We can implement this ourselves fairly "easily" without having to rely on many additional dependencies. All that needs to be done is build APIs for launching the chromium-based web browser, the CDP websocket connection and session management, serialization/deserialization classes+methods for the devtools protocol, and finally a CDP client which plugins can use to write extraction logic.

On Streamlink's Matrix/Gitter channel I've already announced yesterday that I have the whole thing already working for the Twitch plugin, but in a very early stage without any tests and without proper separation in the published git branch. That's what I'm going to finish now.

## Roadmap

1. **CDP spec implementation**

   - [x] #5381

   Add `script/generate-cdp.py` which parses the JSON data of the CDP specification and which then generates {,de}serialization classes and methods for CDP commands and events with proper typing data. Those generated python modules will be committed into the code base and can be updated if needed.

   I will use a modification of the generator script of the [`python-chrome-devtools-protocol`](https://github.com/HyperionGray/python-chrome-devtools-protocol) project, which is MIT licensed, so no problem at all. It works beautifully with my applied fixes and added features.

2. **Implement web browser launcher**

   - [x] #5386 

   This will be based on [`trio`](https://github.com/python-trio/trio) for async I/O code and its async subprocess handling.

   Not writing async code and instead using different threads would make this a thousand times more difficult, especially when writing tests. Compared to `asyncio` of the stdlib, `trio` handles `KeyboardInterrupt`s gracefully and the code is much easier to understand, maintain and test. It's even worth making use of `trio` in other parts of the code at some point in the future, too.

   `trio` is packaged on all major Linux distros.

   The web browser launcher is split into a generic base class and an implementation for Chromium which the CDP client will rely on. The important stuff here is the collection of Chromium executable names, fallback paths and launch arguments.

3. **Implement CDP connection logic and CDP session management**

   - [x] #5388

   Based on [`trio-websocket`](https://github.com/HyperionGray/trio-websocket), which is a bit awkward considering that we already have `websocket-client` as a dependency for the already existing `plugin.api.websocket` implementation. But as said, not writing async code would be insane.

   `trio-websocket` seems to be only partially packaged on Linux. On Arch, it's in the AUR, and Fedora doesn't have an RPM. We can make the whole thing optional though if there are any packaging concerns.

   The CDP connection and session logic will be based on the design of the implementation of `trio-chrome-devtools-protocol` (MIT), with more improvements and fixes.

4. **Implement a CDP client**

   - [x] #5410

   A class with convenience methods for intercepting network requests/responses and for listening to console API calls (for data communication with injected scripts).

   Plugins like Twitch can then set up network hooks for intercepting or reading data, as well as executing scripts. Not going to track this implementation in this thread here.

Further changes are updates to the CLI argparser and Streamlink session for setting/overriding webbrowser paths, CDP host/port, etc.

I hope that I will have everything ready until Twitch turns the requirement on again.

## Concerns

- **Maintenance burden**
  I personally don't care because I'm implementing and will very likely be maintaining this myself anyway.
  The only added dependencies will be `trio`, `trio-websocket`, `typing-extensions` and `pytest-trio` (as well as `inflection` for the generator script, which is irrelevant for packagers). As mentioned above, we can make everything optional, so if a distributor doesn't have those packages, their users won't be able to use any of this.
- **Having the CDP classes in the code base will bloat it up**
  Unfortunate requirement, but I will limit this only to the needed/relevant CDP domains.
- **Chromium-based browser limitation due to the webdriver interface**
  Unfortunate restriction. I explained this above and in the linked comment.
- **Headless mode**
  ~~Since certain sites like Twitch also detect the browser when it's run in headless mode, this will make using such plugins without a desktop environment (loosely speaking) impossible. That is nothing which we can fix though unless we can circumvent such detections, which might be impossible ("canvas fingerprinting", etc.). There are lots of projects on the web which try to achieve this. I'm unaware of any successful ones.~~
   Headless mode apparently does work. We can add a switch for this if it causes problems.
- **User-friendliness**
  Having to launch the web browser which will display its window for a split second if not run in headless mode can be annoying, but once again, nothing which we can fix if headless mode gets detected on some sites. If other sites don't detect headless mode, then there won't be any distraction/disruption. The browser window size can be shrinked to its minimum. There's no way to launch Chromium in minimized mode, but maybe it's possible to force a certain window title, so users can apply rules to their window manager, etc.

**Comment by Hakkin** [comment]
> Since certain sites like Twitch also detect the browser when it's run in headless mode, this will make using such plugins without a desktop environment (loosely speaking) impossible.

I've actually confirmed that the [new Chrome headless](https://developer.chrome.com/articles/new-headless/) introduced in Chrome 112 (with the `--headless=new` flag) currently seems to be undetected by Kasada/Twitch. Who knows if it will stay this way, but since the new headless is based on the same code base as the base browser, I imagine it should be much harder to detect compared to previous headless versions.

**Comment by bastimeyer** [comment]
Tasks 1 and 3 are now complete.

As mentioned on Matrix/Gitter, I have a branch with all the changes in a single unorganized commit with unfinished stuff which I'm updating occasionally. This can be tested with the Twitch plugin.
https://github.com/streamlink/streamlink/compare/master...bastimeyer:streamlink:cdp-wip

----

> Confirmed I can get a `"is_bad_bot": false` using `--headless=new` with remote debugger.

I swear this wasn't working during all my tests, but I checked it again, and appears to be working now. Tested with the current state of my linked branch.

As mentioned in #5386, I will add an option to the `ChromiumWebbrowser` class once everything's done and also a session option / CLI argument, so this can be toggled on and off if it causes problems.

**Comment by bastimeyer** [comment]
Why don't you have a look at the stuff yourself before writing such a comment? All this achieves is forcing me to write another lengthy answer which ultimately wastes my time.

I am not hacking together a temp solution, I am working on a (stable) API which Streamlink plugins can use to launch a web browser in order to extract arbitrary data from websites, similar to existing browser automation tools like Selenium, Puppeteer, etc. The advantage is absolutely clear: we don't have to depend on any re-implementations of JS code in Python, especially if those would be questionable, unreasonable, or very difficult to write and maintain, if even possible at all.

I don't know why you are talking about "WIP browser emulators". Emulation would be trying to extract data from HTTP responses and then either re-implementing (parts of) the JS code in Python or trying to run the stuff in a JS engine written in Python. This isn't just massively prone to errors, but it also requires that everything which modern browsers support has to be supported here too, including all the available web APIs, with reasonable performance and without any security concerns. Not to mention that those scripts are designed to detect automation tools, so they are checking for tons of stuff.

The web browser which gets launched is the same exact version which the user would be using when visiting the site regularly, so this guarantees that we're compatible. All we have to do is intercept network requests via the respective debugging interfaces the browser provides.

This webbrowser API is supposed to be used as a last resort solution and it unfortunately has become a necessity. Enjoy while the requirements are currently disabled again on Twitch.

I have already explained in detail in my comment you've just linked why Twitch's client-integrity check can't be done in Python without doing some insanely crazy stuff. Re-implementing this is a game of cat and mouse you're absolutely not going to win. Remember what the goal of their client-integrity token is.

----

Offtopic:

If you think otherwise, then be my guest, go ahead and figure out the right `/tl` POST body payload which the following script generates when the required cookies are not yet set. The rest has already been figured out and could be re-implemented in Python (the KPSDK message content when the cookies are set by the `/tl` request).

> https://k.twitchcdn.net/149e9513-01fa-4fb0-aad4-566afd725d1b/2d206a39-8ed7-437e-a3be-862e0f06eea3/fp

Let me be so kind and give you some hints:

1. Find the right encoded VM bytecode in the obfuscated JS code (it varies on every request - about 250KiB in size)
2. Find the used alphabet of the encoded bytecode and decode it using the correct algorithm from the obfuscated JS code
3. Understand how the VM in the obfuscated JS code works and which instructions it interprets and what is considered data in the bytecode (AFAICT there are 60 different instructions)
4. Figure out which checks it performs in the browser's environment when it executes the bytecode, so everything can be spoofed. This is what I observed while stepping through it for a bit and while working on the webbrowser API and Twitch plugin implementation:
   - It checks `Navigator.prototype.webdriver`
   - It checks the user agent in various different ways
   - It does some timing related checks
   - It performs lots of other checks, most likely DOM and other web API related stuff for validating the web browser (it even detects wrong Chromium versions when the UA is changed via the CDP, so this implies that they are checking for certain features depending on the version)
   - It executes an encryption algorithm with an unknown 128-bit key and 64-bit initialization vector (the key and IV must be part of the payload though)
5. Find out all the steps it does in order to generate the encrypted `/tl` POST body, so you can get the cookies for requesting `/fp` again, and then write a Python implementation for that
6. Parse the `/fp` KPSDK message (I have the Python code for that here stashed in my local git repo), so you can set the right headers on the `/integrity` GQL API endpoint

And after spending a ton of time working on this with lots of sweat and tears, their checks suddenly change because they want to prevent bots accessing the streams, so all your efforts will be worthless and you'll have anew. A brilliant idea re-implementing this! :brain:

**Comment by Hakkin** [comment]
How does the current implementation work with cookies/headers? Would it use whatever cookies are present in the current chrome session, or does streamlink pass in it's own? I think some people are probably passing in their OAuth tokens to streamlink to bypass ads, so if there is a mismatch between the browser and streamlink the token will be rejected (`user_id` is verified in the token).

**Comment by bastimeyer** [comment]
The `--user-data-dir` param gets set with a temporary directory, which is a requirement anyway, because a new Chromium instance needs to be launched with the remote debugging interface. That wouldn't be the case if the user's default user data dir was used and an instance of Chromium was already running.

See https://github.com/streamlink/streamlink/pull/5386/files#diff-75d6def09629aec8c6ee9a436c7bd9b58545e4ceca3ce8713872f33cd16ec9b9R144
and https://github.com/streamlink/streamlink/pull/5386/files#diff-75d6def09629aec8c6ee9a436c7bd9b58545e4ceca3ce8713872f33cd16ec9b9R54-R123

Regarding auth data, this can be passed to the CDP client, but that part isn't finished yet.

Unfinished work on top of #5388:
https://github.com/streamlink/streamlink/compare/bastimeyer:webbrowser/cdp/init...bastimeyer:streamlink:cdp-wip

**Comment by Hakkin** [comment]
Was also able to confirm Firefox's headless mode is undetected by Kasada, though like the OP issue states, it seems their CDP implementation it too incomplete to be compatible with how streamlink intercepts the token. It would likely be possible using Firefox's own [RDP](https://firefox-source-docs.mozilla.org/devtools/backend/protocol.html), but obviously that's a completely different implementation. Rather than doing anything specific for just Firefox, it would probably be better to wait for [WebDriver BiDi](https://developer.chrome.com/articles/webdriver-bidi/) to be more widely supported and then implement that, it would allow more widespread browser support since pretty much all the major browser vendors have pledged to support it.

If you want to test out Firefox headless, I used the following process
<details>

Create a directory for a profile, add a `user.js` file inside the profile directory with the following options:

```js
user_pref("devtools.chrome.enabled", true);
user_pref("devtools.debugger.prompt-connection", false);
user_pref("devtools.debugger.remote-enabled", true);
```

then you can start Firefox headless
```shell
$ firefox --no-remote --headless --profile "/path/to/profile" --start-debugger-server 6000 https://hakkin.github.io/twitchintegrity/
```

You can interact with the remote debugger through a normal Firefox instance at `about:debugging`, you have to add a new network target at `localhost:6000`. Inspecting the HTML shows a `"is_bad_bot": "false"` response.

If you want to test out Firefox's CDP implementation, the command line option for that is `--remote-debugging-port`, it seems you also need these addition options in `user.js` for it to work properly:

```js
user_pref("fission.bfcacheInParent", false);
user_pref("fission.webContentIsolationStrategy", 0);
```
</details>

**Comment by bastimeyer** [comment]
I've submitted the final PR now, which means the basic webbrowser API stuff is now finished. Certain details like proxy support will be added later. I don't see this as important right now. Both Chromium and Streamlink (requests) support the `HTTP_PROXY`/`HTTPS_PROXY` env vars, so those can be set instead.

As said, I didn't want to merge the PRs individually before everything was ready. However, since we're about to drop py37, we should do that first, maybe also all the other PRs with breaking changes first. Then I can remove py37 compatibility imports from the webbrowser implementation which would require another PR otherwise.

**Comment by bastimeyer** [comment]
Final implementation with the updated Twitch plugin is still on my [`cdp-wip`](https://github.com/bastimeyer/streamlink/tree/cdp-wip) branch, if anyone wants to test this.

Diff compared to the latest submitted changes here:
https://github.com/bastimeyer/streamlink/compare/webbrowser/cdp/client...cdp-wip

**Comment by Hakkin** [comment]
```
34:47.576823][plugins.twitch][debug] Getting live HLS streams for valorant_jpn
34:47.576823][plugins.twitch][info] Acquiring new client-integrity token
34:48.013882][webbrowser.webbrowser][info] Launching web browser: C:\Program Files\Google\Chrome\Application\chrome.exe
34:50.666991][webbrowser.webbrowser][debug] Waiting for web browser process to terminate
34:51.265148][plugins.twitch][debug] Client-Integrity token: {"client_id":"kimne78kx3ncx6brgo4mv6wki5h1ko","client_ip":"xx","device_id":"371acf3e7ba33a43","exp":"2023-07-04T01:34:50Z","iat":"2023-07-03T09:34:50Z","is_bad_bot":"false","iss":"Twitch Client Integrity","nbf":"2023-07-03T09:34:50Z","user_id":""}
34:51.265148][plugins.twitch][info] Is bad bot? False
```

Working nicely. It seems the code to cache the token is commented out though, so it's fetching a new one every time. Is that intentional? Either way it's pretty quick even without caching, and with headless it's more or less completely unnoticeable for the end user.

**Comment by bastimeyer** [comment]
That line is left from an earlier test. The client-integrity token can't be cached unfortunately.

> [plugins.twitch][info] Using cached client-integrity token
> error: No playable streams found on this URL: twitch.tv/...

After #5410 gets merged (may require some fine-tuning, otherwise it's ready), then I will have a proper look at the Twitch plugin.

I doubt though that caching the `/tl` cookies and then parsing the KPSDK message in Python for getting the integrity token makes sense initially. This can be done later as an enhancement.

The important part right now is getting the base implementation ready. This means trying to get an access token first, and if that fails, acquiring a client-integrity token and getting the access token again. The implementation on the `cdp-wip` branch currently always gets a client-integrity token.

**Comment by bastimeyer** [comment]
Btw, I did some more tests with other browsers after adding headless mode to the ChromiumWebbrowser class.

Vivaldi didn't behave 100% correctly and still showed a white box of some kind, but it worked. It also tried to update itself. I did this in my Windows VM, because I didn't want to pollute my host system. There don't seem to be any launch arguments for disabling that behavior. Typical Windows stuff.

Edge and Brave worked as intended.

Any other worthy Chromium-based browsers that can be tested?

**Comment by Hakkin** [comment]
> That line is left from an earlier test. The client-integrity token can't be cached unfortunately.

I believe it's not working for 2 reasons:
There was a bug in the code,
```python
client_integrity = self.cache.get(self._CACHE_KEY_CLIENT_INTEGRITY)
```
should be
```python
client_integrity, expiration = self.cache.get(self._CACHE_KEY_CLIENT_INTEGRITY)
```
(this fixes it currently),
but another issue is that the Device-ID used to fetch the token is randomized each time and not saved, currently Twitch has stopped verifying this, so it works for now, but when the token restriction was in place they were, so I imagine whenever it comes back this will become an issue again.

**Comment by Hakkin** [comment]
> ```python
> client_integrity, expiration = self.cache.get(self._CACHE_KEY_CLIENT_INTEGRITY)
> ```

Oops, this actually also doesn't work if there's not already a cached value. I guess it should be something like
```python
client_integrity_cached = self.cache.get(self._CACHE_KEY_CLIENT_INTEGRITY)
if client_integrity_cached:
  client_integrity, expiration = client_integrity_cached
  log.info(f"Using cached client-integrity token {client_integrity}, expires at {expiration}")
```

---

> Any other worthy Chromium-based browsers that can be tested?

I tried Opera GX since it's commonly shilled amongst the Twitch audience, but just setting the executable path doesn't work. If I manually start the `launcher.exe` with `--remote-debugging-port=9222` then manually pass the port to streamlink, it does work, so it does seem to recognize Chrome command line flags. The streamlink error is 
```shell
[webbrowser.webbrowser][warning] Web browser process ended early
```

So it might be some kind of issue with it immediately spinning off from a launcher process (trying to pass in either `launcher.exe` or `opera.exe` results in the same behavior).

**Comment by bastimeyer** [comment]
I've deleted the `cdp-wip` branch and submitted #5414. Please give this a try. Thanks.

Let me close this thread now, since the first iteration of the webbrowser API is now fully implemented on the master branch.
